src/modules/withdraw-queue/WithdrawQueue.sol

```diff
@@ -0,0 +1,301 @@
+            if (request.sharesToWithdraw == 0) revert WithdrawQueue__NoShares();
+
+            // User gets whatever their execution share price is.
+            requiredAssets += _calculateAssetAmount(
+                request.sharesToWithdraw,
+                request.executionSharePrice,
+                solveData.shareDecimals
+            );
+
+            // If all checks above passed, the users request is valid and should be fulfilled.
+            sharesToSolver += request.sharesToWithdraw;
+            request.inSolve = true;
+            // Transfer shares from user to solver.
+            share.safeTransferFrom(users[i], solver, request.sharesToWithdraw);
+        }
+
+        ISolver(solver).finishSolve(runData, sharesToSolver, requiredAssets);
+
+        for (uint256 i; i < users.length; ++i) {
+            WithdrawRequest storage request = userWithdrawRequest[users[i]][share];
+
+            if (request.inSolve) {
+                // We know that the minimum price and deadline arguments are satisfied since this can only be true if they were.
+
+                // Send user their share of assets.
+                uint256 assetsToUser = _calculateAssetAmount(
+                    request.sharesToWithdraw,
+                    request.executionSharePrice,
+                    solveData.shareDecimals
+                );
+
+                solveData.asset.safeTransferFrom(solver, users[i], assetsToUser);
+
+                emit RequestFulfilled(
+                    users[i],
+                    address(share),
+                    request.sharesToWithdraw,
+                    assetsToUser,
+                    block.timestamp
+                );
+
+                // Set shares to withdraw to 0.
+                request.sharesToWithdraw = 0;
+                request.inSolve = false;
+            } else revert WithdrawQueue__UserNotInSolve();
+        }
+    }
+
+    /**
+     * @notice Helper function solvers can use to determine if users are solvable, and the required amounts to do so.
+     * @notice Repeated users are not accounted for in this setup, so if solvers have repeat users in their `users`
+     *         array the results can be wrong.
+     */
+    function viewSolveMetaData(
+        ERC4626 share,
+        address[] calldata users
+    ) external view returns (SolveMetaData[] memory metaData, uint256 totalRequiredAssets, uint256 totalSharesToSolve) {
+        // Get Solve Data.
+        SolveData memory solveData;
+        solveData.asset = share.asset();
+        solveData.assetDecimals = solveData.asset.decimals();
+        solveData.shareDecimals = share.decimals();
+
+        // Setup meta data.
+        metaData = new SolveMetaData[](users.length);
+
+        for (uint256 i; i < users.length; ++i) {
+            WithdrawRequest memory request = userWithdrawRequest[users[i]][share];
+
+            metaData[i].user = users[i];
+
+            if (block.timestamp > request.deadline) {
```

Comment:
Let's walk through this together, just want to make sure I understand the flags and solver process.